% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/estimate.R
\name{drseqCount}
\alias{drseqCount}
\title{Construct DrSeq Data}
\usage{
drseqCount(
  event,
  sampleData,
  design = ~sample + exon + condition:exon,
  remove.overlap.exon = FALSE,
  anno.prefix = "drseq.annotation",
  anno.format = "gff2",
  minMQS = 10,
  isPairedEnd = TRUE,
  cores = max(1, detectCores() - 2),
  verbose = FALSE,
  ...
)
}
\arguments{
\item{event}{a \code{surf} object from \link{parseEvent}.}

\item{sampleData}{\code{data.frame}, describes the RNA-seq samples and
contains at least two columns -- \code{bam} and \code{condition},
whose \code{row.names} represent sample names.}

\item{design}{
    A formula which specifies the design of the experiment. It must
    specify an interaction term between a variable from the sampleData
    columns with the 'exon' variable. By default, the design 
    will be '~ sample + exon + condition:exon'. This formula indicates
    the contrast between 'condition' and exon', i.e. differences in exon 
    usage due to changes in the 'condition' variable.  See the vignette 
    for more examples of other designs.
  }

\item{remove.overlap.exon}{\code{logical}, remove overlapping exons across genes
(default to \code{FALSE}).}

\item{anno.prefix}{\code{character}, file names for outputting annotation files.
If prefix is absent, hidden file will be output to the current working
directory.}

\item{anno.format}{\code{character}, e.g. "gtf", as accepted by
\link[rtracklayer:export]{rtracklayer::export}.}

\item{minMQS, isPairedEnd}{as defined in \link[Rsubread:featureCounts]{Rsubread::featureCounts}.
Note that the default is customized for SURF (see details for more
information).}

\item{cores}{\code{integer}, number of available workers, sent to \code{nthreads} of
\code{featureCounts}.}

\item{verbose}{\code{logical}, whether (\code{TRUE}) to echo progress.}

\item{...}{additional parameters for \link[Rsubread:featureCounts]{Rsubread::featureCounts}.}
}
\value{
a \code{surf} object, with \code{drseqData} slot updated.
}
\description{
This function creates the \code{drseqData} slot needed by DrSeq analysis, which is
a \link[DEXSeq:DEXSeqDataSet]{DEXSeq::DEXSeqDataSet}. This function requires two DrSeq annotation files
created by \link{prepDrseqAnno}. Two files have the same prefix (\code{anno.prefix}),
the same suffix (\code{anno.format}), and only differ by "exon/event".
If annotation files are missing, this function can create them freshly, which
might take some time. This function counts the RNA-seq reads on ATR events
and genes using \code{featureCounts}.
}
\details{
If you used Illumina HiSeq 2000, set \code{strandSpecific = 2} (reversed strand).
}
