% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/estimate.R
\name{faseqCount}
\alias{faseqCount}
\title{Construct FASeq Data Set}
\usage{
faseqCount(
  event,
  sampleData,
  signal.type = "FPKM",
  FUN.aggregate = "mean",
  minMQS = 10,
  minOverlap = 12,
  isPairedEnd = T,
  cores = max(1, detectCores() - 2),
  verbose = F,
  ...
)
}
\arguments{
\item{event}{a \code{surf} object.}

\item{sampleData}{\code{data.frame}, must contain two columns -- "bam" and "condition" (for "IP" and "input", where "IP" should come first), whose \code{row.names} represent the sample names. "bam" is the file name of CLIP-seq bam. "condition" will be coerced to factor, whose first level will be treated as IP, and the second level as input.}

\item{signal.type}{\code{character}, indicate the type of feature signal wanted, support "TPM" for Transcripts Per Kilobase Million, "FPKM" for Fragments Per Kilobase Million (for paired-end reads) and Reads Per Kilobase Million (for single-end reads), and "raw.count" for raw.count read counts}

\item{FUN.aggregate}{\code{function}, used for aggregating signals within \code{condition}, default to mean.}

\item{cores}{\code{integer}, number of available workers, sent to \code{nthreads} of \link{featureCounts}}

\item{verbose}{\code{logical}, whether (default to \code{TRUE}) to echo progress}

\item{minMQS;minOverlap;isPairedEnd;...}{parameters for \link{featureCounts}. \code{minMQS} is default to 10, and \code{minOverlap} is default to 12 (25\% of the typical read length of CLIP-seq (~50bp)), and \code{isPairedEnd} is default to \code{TRUE}.}
}
\value{
a \code{surf} object, with (i) one column \code{featureSignal} added, (2) \code{faseqData} slot updated, and (3) \code{sampleData} slot updated.
}
\description{
This function quantifies feature signals for location features using CLIP-seq data.
You align CLIP-seq reads to the genome and provide FASeq with the resulting bam files.
We will take care of the rest.
}
\details{
If your sequencing platform is Illumina HiSeq 2000, set \code{strandSpecific = 2}.
}
\references{
\url{https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/}
}
\keyword{CLIP-seq}
\keyword{feature}
\keyword{signal,}
