---
title: "Intergrative analysis of CLIP-seq and RNA-seq with `surf`"
author: "Fan Chen (fan.chen@wisc.edu)"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{Intergrative analysis of CLIP-seq and RNA-seq with `surf`}
  %\usepackage[UTF-8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      tidy = TRUE, 
                      tidy.opts = list(comment = FALSE))
library(surf)
```

```{r prepare mm annotation, include=F, eval=F}
library(rtracklayer)
library(usethis)
## parse events from genome annotation (GTF)
## /path/to/genome/annotation/file
anno_file <- "~/Downloads/gencode.vM23.primary_assembly.annotation.gtf"
anno_mm <- import(anno_file)
gene_id <- anno_mm[seqnames(anno_mm) == "chr14" &
                     anno_mm$gene_type == "protein_coding" &
                     anno_mm$type == "gene"]$gene_id
gene_id_sampled <- sample(unique(gene_id), 50)
anno_mm_select <- anno_mm[anno_mm$gene_id %in% gene_id_sampled]

## Mus musculus
export(anno_mm_select, "~/Downloads/gencode.vM23.primary.example.gtf")
```

```{r prepare hs annotation, include=F, eval=F}
library(rtracklayer)
library(usethis)
## parse events from genome annotation (GTF)
## /path/to/genome/annotation/file
anno_file <- "~/Downloads/gencode.v32.primary_assembly.annotation.gtf"
anno_hs <- import(anno_file)
gene_id <- anno_hs[seqnames(anno_hs) == "chr16" &
                     anno_hs$gene_type == "protein_coding" &
                     anno_hs$type == "gene"]$gene_id
gene_id_sampled <- sample(unique(gene_id), 24)
anno_hs_select <- anno_hs[anno_hs$gene_id %in% gene_id_sampled]

## Homo sapiens
export(anno_hs_select, "~/Downloads/gencode.v32.primary.example.gtf")

## export gene ranges 
genes <- reduce(anno_hs_select)
strand(genes) <- "*"
genes <- resize(genes, width(genes) + 300, fix = "center")
export(genes, "~/Downloads/gencode.v32.primary.region.bed")
```

This document provides an example of using `surf` to ananlyze CLIP-seq and RNA-seq for predicting RNA-binding protein functions. `surf` is currently available from [GitHub](https://github.com/fchen365/surf.git). The analysis consists of four steps, which we describe in order. 

## Step 1: parse ATR events from genome annotation.

The first step is to part genome annotation for alternative transcriptional regulation (ATR) events. For this, `surf` provides a `parseEvent` function which takes as input various types of genome annotation files, e.g., GTF or GFF. 
```{r parse, eval=F}
## parse events from genome annotation (GTF)
event <- parseEvent("~/Downloads/gencode.v32.primary.example.gtf")
```
```{r, echo=F}
# saveRDS(event, "~/Downloads/intermediate1.rds")
event <- readRDS("~/Downloads/intermediate1.rds")
```

The function allows multiple options. 

* `cores` specifies the number of computing processors. 
* `min.event.length` allows you to filter out ATR event by a lower limit of lenght (in bp).
* `location.feature` determines whether location features should also be extracted. This is useful if you are interested in perform differential regulation of ATR event using RNA-seq (DrSeq). The location features will be used to quantify CLIP-seq signals (and will not be used by the second step, DrSeq). 
* `depth.exon` and `depth.intron` is relavent only if `location.feature=TRUE`. The configure the the sizes of location features. 
* `remove.duplicate` defines wehther identical event should be removed. 
* `verbose` controls whether or not the progress should be print out. 

This step outputs a `surf` object. The `surf` object contains a `DataFrame`. 
```{r}
event
```

With `mcols()` function, we can inspect the description of each column.
```{r}
mcols(event)
```

The output `surf` object also comes with a `genePartsList` slot, for gene parts list. `genePartsList` is a `DataFrame` of 5 columns. Use `mcols()` to inspect the descriptions of columns.
```{r}
pl <- genePartsList(event)
pl
```
```{r}
mcols(pl)
```

For `surf` object, the useful genomic function, `findOverlaps`, are implemented. 
```{r}
gr0 <- GRanges(seqnames = Rle("chr16"), 
               IRanges(89710000, width = 10000))
findOverlaps(event, subject = gr0)
```
Subsetting by overlaps is also implemented (`subsetByOverlaps`).
```{r}
subsetByOverlaps(event, ranges = gr0)
```


## Step 2: detect differential regulation (DR) of ATR events

```{bash prepare RNA-seq bam files, include=F, eval=F}
## shRNA-seq, target: CPSF6
DIR=rna-seq/bam
KD1=$DIR/ENCSR384BDV/ENCFF291ERF
KD2=$DIR/ENCSR384BDV/ENCFF956SFY
WT1=$DIR/ENCSR424QCW/ENCFF529TIM
WT2=$DIR/ENCSR424QCW/ENCFF361KUQ
# REGION=~/Downloads/gencode.v32.primary.region.bed
module load python/miniconda
conda activate /workspace/fanchen/conda2
samtools view -b $KD1.sorted.bam chr16 -h chr16 > $KD1.chr16.bam
samtools view -b $KD2.sorted.bam chr16 -h chr16 > $KD2.chr16.bam
samtools view -b $WT1.sorted.bam chr16 -h chr16 > $WT1.chr16.bam
samtools view -b $WT2.sorted.bam chr16 -h chr16 > $WT2.chr16.bam
conda deactivate
```

To detect the differnetial regulation of ATR events, use `drseq()` function. The function requires a `sampleData` table that specifies the sample information of RNA-seq. In particular, a `condition` column indicating the experimental conditions is required, as well as a `bam` column giving the file directory for the aligned RNA-seq reads. 
```{r drseq, eval=F}
rna_seq_sample <- data.frame(
  row.names = c('sample1', 'sample2', 'sample3', 'sample4'),
  bam = paste0("~/Downloads/",c("KD1", "KD2", "WT1", "WT2"),".bam"),
  condition = c('knockdown', 'knockdown', 'wildtype', 'wildtype'),
  stringsAsFactors = FALSE
) 
event <- drseq(event, rna_seq_sample)
```
```{r, echo=F}
# saveRDS(event, "~/Downloads/intermediate2.rds")
event <- readRDS("~/Downloads/intermediate2.rds")
```

The output of `drseq` (still a `surf` object) contains 6 additional columns (the 7-12th) to the existing output from Step 1. 
```{r}
event[,7:12]
```
Again, you can use mcols() to check the description of each columns.
```{r}
mcols(event)[7:12,]
```

In addition to the main output, the output also contains a `drseqResults` slot, which can be simply accessed by the `drseqResults()` function. 
```{r}
drr <- drseqResults(event)
drr
```
The descriptions of columns can be found by the `mcols()` function:
```{r}
mcols(drr)
```

For both `surf` and `drseqResults`, two visualization methods are available: (1) `plotDispFunc()` which plots the fitted dispersion functions for each ATR event type, and (2) `volcano.plot()` which plots the volcano plot for each ATR events stratified by different event types. 


## Step 3: test the functional association between RBP binding and DR events. 

```{bash prepare CLIP-seq bam files, include=F, eval=F}
## eCLIP-seq, target: CPSF6
DIR=clip-seq/bam
IP1=$DIR/ENCSR532VUB/ENCFF751GZQ
IP2=$DIR/ENCSR532VUB/ENCFF550ZSZ
SMI=$DIR/ENCSR249WNK/ENCFF842FFD
# REGION=~/Downloads/gencode.v32.primary.region.bed
module load python/miniconda
conda activate /workspace/fanchen/conda2
samtools sort $IP1.bam -o $IP1.sort.bam
samtools index $IP1.sort.bam
samtools view -b $IP1.sort.bam chr16 -h chr16 > $IP1.chr16.bam
samtools sort $IP2.bam -o $IP2.sort.bam
samtools index $IP2.sort.bam
samtools view -b $IP2.sort.bam chr16 -h chr16 > $IP2.chr16.bam
samtools sort $SMI.bam -o $SMI.sort.bam
samtools index $SMI.sort.bam
samtools view -b $SMI.sort.bam chr16 -h chr16 > $SMI.chr16.bam
conda deactivate
```

To detect the differnetial regulation of ATR events, use `faseq()` function. The function requires a `sampleData` table that specifies the sample information of CLIP-seq. In particular, a `condition` column indicating the experimental conditions is required, as well as a `bam` column giving the file directory for the aligned CLIP-seq reads. In this example, the `min.size` parameter is set to 3 for presentation purpose. For reliable statistical inference, we recommend a value of at least 100 (default, or 50 for event types like retained intron (RI)) in genome-wide analysis. `fdr.cutoff` is the cutoff of FDR (BH adjusted p-values) in functional association testings. The default is 0.05 (here we set this to 0.3 for illustration purpose). `signal.cutoff` is the cut-off threshold for the eCLIP signals, default to 20, and is set to 2 for illustration purpose. 
```{r, eval=F}
clip_seq_sample = data.frame(
  row.names = c('sample5', 'sample6', 'sample7'),
  bam = paste0("~/Downloads/",c("IP1", "IP2", "SMI"),".bam"),
  condition = c('IP', 'IP', 'SMI'),
  stringsAsFactors = FALSE
)
event <- faseq(event, clip_seq_sample, 
               min.size = 3, fdr.cutoff = 0.3, signal.cutoff = 2)
```
```{r, echo=F}
# saveRDS(event, "~/Downloads/intermediate3.rds")
event <- readRDS("~/Downloads/intermediate3.rds")
```

The output of `faseq` (still a `surf` object) adds two additional columns (the 13th and 14th) to the existing output from Step 2. 
```{r}
event[,13:14]
```
Again, you can use mcols() to check the description of each columns.
```{r}
mcols(event)[13:14,]
```

In addition to the main output, the output also contains a `faseqResults` slot, which can be simply accessed by the `faseqResults()` function. 
```{r}
far <- faseqResults(event)
far
```
The descriptions of columns can be found by the `mcols()` function:
```{r}
mcols(far)
```

Two visualization methods are available: (1) `fa.plot()` which plots the fitted dispersion functions for each ATR event type, and (2) `inferredFeature()` which plots the volcano plot for each ATR events stratified by different event types. 

```{r, fig.width=8}
fa.plot(event)
```
```{r}
inferredFeature(event)
```


## Step 4: RBP-related discovery using external transcriptome data

`daseq()` function performs RBP-related discovery. The function requires a `sampleData` table that specifies the sample information of CLIP-seq. In particular, a `condition` column indicating the experimental conditions is required, as well as a `bam` column giving the file directory for the aligned CLIP-seq reads. In this example, the `min.size` parameter is set to 3 for presentation purpose. For reliable statistical inference, we recommend a value of at least 100 (default, or 50 for event types like retained intron (RI)) in genome-wide analysis.
```{r, eval=F}
## rank transcripts (TPM)
exprMat <- readRDS('~/Downloads/TcgaTargetGtex_rsem_isoform_tpm_laml_blood_10each.rds')
## sample data
ext_sample <- data.frame(
  condition = rep(c('TCGA', 'GTEx'), each = 10),
  row.names = colnames(exprMat)
)

## differential activity (transcript)
event <- daseq(event, getRankings(exprMat), cores = 1, ext_sample)
```
```{r, echo=F}
# saveRDS(event, "~/Downloads/intermediate4.rds")
event <- readRDS("~/Downloads/intermediate4.rds")
```

The output also contains a `daseqResults` slot, which can be simply accessed by the `daseqResults()` function. 
```{r}
dar <- daseqResults(event)
dar
```
The descriptions of columns can be found by the `mcols()` function:
```{r}
mcols(dar)
```
